engine/src/main/java/org/hibernate/validator/internal/metadata/core/ConstraintHelper.java

```diff
@@ -296,9 +297,13 @@
 	private static final String JODA_TIME_CLASS_NAME = "org.joda.time.ReadableInstant";
 	private static final String JAVA_MONEY_CLASS_NAME = "javax.money.MonetaryAmount";
 
-	// immutable
+	@Immutable
 	private final Map<Class<? extends Annotation>, List<? extends ConstraintValidatorDescriptor<?>>> builtinConstraints;
 
+	private final ConcurrentHashMap<Class<? extends Annotation>, Boolean> externalConstraints = new ConcurrentHashMap<>();
```

Comment:
More generally, though, it seems to go a bit against the goal of reducing memory usage.